C-Plus-Plus/cp/zalgorithm.cpp

```diff
@@ -0,0 +1,58 @@
+#include <bits/stdc++.h>
+using namespace std;
+
+void getZ(string s, int z[]) {
+    int n = s.size();
+    int left = 0, right = 0;
+    for (int i = 1; i < n; ++i) {
+        if(i > right) {
+            left = right = i;
+            while (right < n && s[right-left] == s[right])
+                right++;
+            z[i] = right-left;
+            right--;
+        }
+        else {
+            int k = i-left;
+            if(z[k] < right-i+1) {
+                z[i] = z[k];
+            }
+            else {
+                left = i;
+                while (right < n && s[right-left] == s[right])
+                    right++;
+                z[i] = right-left;
+                right--;
+            }
+        }
+    }
+}
+
+int main(){
+    string text, pattern;
+    cin >> text >> pattern;
```

Comment:
Prompt user to input the original string and the pattern to be searched separately.